### PR TITLE
PE-2618: Sync from Snapshots

### PR DESCRIPTION
**Pull request by matibat**
Integration branch for the Story's sub-tasks.

--- Releases ---
Android release: https://appdistribution.firebase.google.com/testerapps/1:305132849030:android:6cf0cd5ec064fad3ffce07/releases/7unj2boou3560
iOS release: https://appdistribution.firebase.google.com/testerapps/1:305132849030:ios:06ea33a95a2e0d42ffce07/releases/56ec6b5bfal0g

### CODE REVIEW

File: lib/blocs/sync/utils/parse_drive_transactions.dart [hunk 1]
```diff
@@ -42,15 +44,20 @@ Stream<double> _parseDriveTransactionsIntoDatabaseEntities({
   final owner = await arweave.getOwnerForDriveEntityWithId(drive.id);
 
   yield* _batchProcess<
-          DriveEntityHistory$Query$TransactionConnection$TransactionEdge>(
+          DriveEntityHistory$Query$TransactionConnection$TransactionEdge$Transaction>(
```
[the remark this answers is not in the document]
Comment (matibat) [reply lib/blocs/sync/utils/parse_drive_transactions.dart]: PE-2782 🚀 
359384fbefd7182031a7174efa6b9d71a48a9e74

File: lib/utils/snapshots/snapshot_item.dart [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+      getNextStream() {
+    _currentIndex++;
+    if (currentIndex >= subRanges.rangeSegments.length) {
+      throw SubRangeIndexOverflow(index: currentIndex);
+    }
+
+    return _getNextStream();
+  }
+
+  Stream<DriveEntityHistory$Query$TransactionConnection$TransactionEdge$Transaction>
```
[the remark this answers is not in the document]
Comment (matibat) [reply lib/utils/snapshots/snapshot_item.dart]: There you go, Javed 😭 - 359384fbefd7182031a7174efa6b9d71a48a9e74

File: lib/blocs/sync/utils/sync_drive.dart [hunk 3]
```diff
@@ -37,16 +37,57 @@ Stream<double> _syncDrive(
...
+  final transactions = <
+      DriveEntityHistory$Query$TransactionConnection$TransactionEdge$Transaction>[];
+
+  final snapshotsStream = arweaveService
+      .getAllSnapshotsOfDrive(
+        driveId,
+        lastBlockHeight,
+      )
+      .map(
+        (edge) => edge.node,
```
[the remark this answers is not in the document]
Comment (matibat) [reply lib/blocs/sync/utils/sync_drive.dart]: Much cleaner :) - ae51d75bceeea90b8a9d34c3b1392a1fb2793dae

File: lib/utils/snapshots/snapshot_item.dart [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+        print('Error while fetching Snapshot Data - $e');
+      },
+    );
+
+    final dataBytesAsString = String.fromCharCodes(dataBytes.data);
+    return _cachedSource = dataBytesAsString;
+  }
+
+  String get _dataUri {
+    return 'https://arweave.net/$txId';
```
[the remark this answers is not in the document]
Comment (matibat) [reply lib/utils/snapshots/snapshot_item.dart]: 04ca8a0c97fee47cfb22741fce39c760d6acd9a5

**Review by karlprieb — APPROVED** [review]

**Pull request merged**